fossas/spectrometer
# fix(go): Correctly parse and handle pseudo-versions of Go modules

Fixes https://github.com/fossas/issues/issues/474.

I was investigating https://github.com/fossas/issues/issues/467 when I discovered that `fixVersion` was intentionally handling pseudo-versions.

Unfortunately, this parsing is wrong - it sometimes parses versions that are _not_ pseudo-versions as if they _were_ pseudo-versions. For an example, see https://github.com/fossas/issues/issues/474, reproduced below from https://github.com/hashicorp/nomad/blob/c4a7728e0323ba27ec88d561d8e1c7cdb8c9739a/go.mod#L106:

```
	github.com/opencontainers/runc v1.0.0-rc92
```

And from https://github.com/hashicorp/nomad/blob/c4a7728e0323ba27ec88d561d8e1c7cdb8c9739a/go.mod#L13:

```
	github.com/hashicorp/nomad/api => ./api
```

This PR adds handling for two cases:

- Relative file imports should not be uploaded as dependencies.
- Pre-release semver imports that are not pseudo-versions should be correctly uploaded as full semvers.

Ideally, we would just upload full pseudo-versions as well (instead of trimming to commit hash) which will also preserve tagged versions when available, but this will require additional backend changes (see https://github.com/fossas/issues/issues/473).

## Testing plan

1. Run `fossa analyze -o` on Hashicorp's Nomad project.
2. Check that the file import is correctly excluded.
3. Check that the `runc` prerelease semver version is correctly parsed.
